### PR TITLE
Python ij review

### PR DESCRIPTION
**Pull request by jburel**
* Review the setup 
* Add python script example

To test, the easiest is to follow option 2 and run the python script under ``scripts/python``

cc @pwalczysko

### CODE REVIEW

**Comment by pwalczysko** [comment]
``...omero-guide-fiji/docs/imagej_python.rst:69: WARNING: image file not readable: images/terminal.png``

Image is missing

File: docs/imagej_python.rst [hunk 1]
```diff
@@ -11,32 +11,125 @@ This is much easier that when using the Java API and Bio-Formats plugin.
...
 - Install omero-py and `pyimagej <https://pypi.org/project/pyimagej/>`_::
 
     $ conda install -c conda-forge pyimagej
 
     $ conda install -c ome omero-py
 
+*Option 2*
+~~~~~~~~~~
+
+Create a local Docker Image using ``repo2docker``, see :download:`README.md <https://github.com/ome/omero-guide-python/blob/master/README.md>`::
```
Comment (pwalczysko) [inline docs/imagej_python.rst]: In the README, which is hosted in omero-guides-python is a sentence "clone this repository". But in the sense of this workflow, the omero-guides-python repo does not need to be cloned (note that when you click onto the README, it does not download the file locally, instead, opens it in github)

File: docs/imagej_python.rst [hunk 2]
```diff
@@ -11,32 +11,125 @@ This is much easier that when using the Java API and Bio-Formats plugin.
...
+Step-by-Step
+------------
+
+The script used in this document is :download:`run_macro_python.py <../scripts/python/run_macro_python.py>`.
+One of the advantage of this approach is that we can load only the 2D-planes we wish to analyze.
+
+In this example, Fiji has been installed locally. It is possible to install Fiji "on the fly",
+see `ImageJ Tutorials <https://nbviewer.jupyter.org/github/imagej/tutorials/blob/master/notebooks/ImageJ-Tutorials-and-Demo.ipynb>`_ for more options.
+
+To run the macro, we use ImageJ1 windows so it requires using ImageJ in GUI mode and requires handling the resulting windows. This is the reason the parameter `headless` is set to `False`.
```
Comment (pwalczysko) [inline docs/imagej_python.rst]:
```suggestion
To run the macro, we use ImageJ1 graphical user interface (GUI) elements, and thus it requires using ImageJ in GUI mode. In this GUI mode, the resulting windows content is handled. This is the reason for the parameter `headless` to be set to `False`.
```

Comment (pwalczysko) [reply docs/imagej_python.rst]: I think also this paragraph needs to be nearer to the code block below where you are setting the headless parameter.

**Comment by pwalczysko** [comment]
```
(notebook) pwalczysko@c30f85666085:~/scripts$ ls
crop.ijm  groovy  python
(notebook) pwalczysko@c30f85666085:~/scripts$ cd python/
(notebook) pwalczysko@c30f85666085:~/scripts/python$ ls
run_macro_python.py
(notebook) pwalczysko@c30f85666085:~/scripts/python$ python run_macro_python.py
Host [wss://idr.openmicroscopy.org/omero-ws]:
Image ID [1884807]:
initializing fiji...
Jun 11, 2020 2:22:36 PM java.util.prefs.FileSystemPreferences$1 run
INFO: Created user preferences directory.
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/srv/conda/vnc/Fiji.app/jars/logback-classic-1.2.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/srv/conda/vnc/Fiji.app/plugins/omero_ij-5.5.11-all.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [ch.qos.logback.classic.util.ContextSelectorStaticBinder]
[ERROR] Failed to instantiate IJ1.
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at net.imagej.patcher.LegacyEnvironment.newImageJ1(LegacyEnvironment.java:475)
        at net.imagej.legacy.LegacyService.initialize(LegacyService.java:457)
        at org.scijava.service.ServiceHelper.createServiceRecursively(ServiceHelper.java:354)
        at org.scijava.service.ServiceHelper.createExactService(ServiceHelper.java:270)
        at org.scijava.service.ServiceHelper.loadService(ServiceHelper.java:232)
        at org.scijava.service.ServiceHelper.loadService(ServiceHelper.java:195)
        at org.scijava.service.ServiceHelper.loadServices(ServiceHelper.java:167)
        at org.scijava.Context.<init>(Context.java:281)
        at org.scijava.Context.<init>(Context.java:237)
        at org.scijava.Context.<init>(Context.java:177)
        at org.scijava.Context.<init>(Context.java:163)
        at net.imagej.ImageJ.<init>(ImageJ.java:77)
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at net.imagej.patcher.LegacyEnvironment.newImageJ1(LegacyEnvironment.java:471)
        ... 11 more
Caused by: java.awt.AWTError: Can't connect to X11 window server using ':1.0' as the value of the DISPLAY variable.
        at sun.awt.X11GraphicsEnvironment.initDisplay(Native Method)
        at sun.awt.X11GraphicsEnvironment.access$200(X11GraphicsEnvironment.java:65)
        at sun.awt.X11GraphicsEnvironment$1.run(X11GraphicsEnvironment.java:115)
        at java.security.AccessController.doPrivileged(Native Method)
        at sun.awt.X11GraphicsEnvironment.<clinit>(X11GraphicsEnvironment.java:74)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at java.awt.GraphicsEnvironment.createGE(GraphicsEnvironment.java:103)
        at java.awt.GraphicsEnvironment.getLocalGraphicsEnvironment(GraphicsEnvironment.java:82)
        at java.awt.Window.initGC(Window.java:475)
        at java.awt.Window.init(Window.java:495)
        at java.awt.Window.<init>(Window.java:537)
        at java.awt.Frame.<init>(Frame.java:420)
        at ij.ImageJ.<init>(ImageJ.java:142)
        at ij.ImageJ.<init>(ImageJ.java:130)
        ... 16 more
[ERROR] Exception during event handling:
        [Event] org.scijava.service.event.ServicesLoadedEvent
        context = org.scijava.Context@78b41097
        consumed = false
        [Subscriber] fiji.DefaultFijiService [priority = 0.0]
        [Method] protected void fiji.DefaultFijiService.onEvent(org.scijava.service.event.ServicesLoadedEvent)
java.lang.NoClassDefFoundError: Could not initialize class sun.awt.X11GraphicsEnvironment
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at java.awt.GraphicsEnvironment.createGE(GraphicsEnvironment.java:103)
        at java.awt.GraphicsEnvironment.getLocalGraphicsEnvironment(GraphicsEnvironment.java:82)
        at sun.awt.X11.XToolkit.<clinit>(XToolkit.java:126)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at java.awt.Toolkit$2.run(Toolkit.java:860)
        at java.awt.Toolkit$2.run(Toolkit.java:855)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.awt.Toolkit.getDefaultToolkit(Toolkit.java:854)
        at fiji.gui.FileDialogDecorator.registerAutomaticDecorator(FileDialogDecorator.java:138)
        at fiji.DefaultFijiService.actuallyInitialize(DefaultFijiService.java:33)
        at fiji.DefaultFijiService.onEvent(DefaultFijiService.java:61)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.scijava.event.DefaultEventService$ProxySubscriber.onEvent(DefaultEventService.java:301)
        at org.scijava.event.DefaultEventService$ProxySubscriber.onEvent(DefaultEventService.java:275)
        at org.bushe.swing.event.ThreadSafeEventService.publish(ThreadSafeEventService.java:971)
        at org.scijava.event.DefaultEventBus.access$201(DefaultEventBus.java:57)
        at org.scijava.event.DefaultEventBus$2.run(DefaultEventBus.java:217)
        at org.scijava.thread.DefaultThreadService.lambda$wrap$1(DefaultThreadService.java:214)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
[ERROR] Exception during event handling:
        [Event] org.scijava.service.event.ServicesLoadedEvent
        context = org.scijava.Context@78b41097
        consumed = false
        [Subscriber] sc.fiji.compat.DefaultFijiService [priority = 0.0]
        [Method] protected void sc.fiji.compat.DefaultFijiService.onEvent(org.scijava.service.event.ServicesLoadedEvent)
java.lang.NoClassDefFoundError: Could not initialize class sun.awt.X11.XToolkit
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at java.awt.Toolkit$2.run(Toolkit.java:860)
        at java.awt.Toolkit$2.run(Toolkit.java:855)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.awt.Toolkit.getDefaultToolkit(Toolkit.java:854)
        at sc.fiji.compat.FileDialogDecorator.registerAutomaticDecorator(FileDialogDecorator.java:138)
        at sc.fiji.compat.DefaultFijiService.actuallyInitialize(DefaultFijiService.java:30)
        at sc.fiji.compat.DefaultFijiService.onEvent(DefaultFijiService.java:57)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.scijava.event.DefaultEventService$ProxySubscriber.onEvent(DefaultEventService.java:301)
        at org.scijava.event.DefaultEventService$ProxySubscriber.onEvent(DefaultEventService.java:275)
        at org.bushe.swing.event.ThreadSafeEventService.publish(ThreadSafeEventService.java:971)
        at org.scijava.event.DefaultEventBus.access$201(DefaultEventBus.java:57)
        at org.scijava.event.DefaultEventBus$2.run(DefaultEventBus.java:217)
        at org.scijava.thread.DefaultThreadService.lambda$wrap$1(DefaultThreadService.java:214)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
connecting to IDR...
running the macro...
Traceback (most recent call last):
  File "run_macro_python.py", line 101, in <module>
    main()
  File "run_macro_python.py", line 94, in main
    analyse(ij, conn, image_id)
  File "run_macro_python.py", line 77, in analyse
    ij.ui().show('Image', ij.py.to_java(plane))
  File "jnius/jnius_export_class.pxi", line 1002, in jnius.JavaMultipleMethod.__call__
  File "jnius/jnius_export_class.pxi", line 734, in jnius.JavaMethod.__call__
  File "jnius/jnius_export_class.pxi", line 828, in jnius.JavaMethod.call_method
  File "jnius/jnius_utils.pxi", line 93, in jnius.check_exception
jnius.JavaException: JVM exception occurred: Could not initialize class sun.awt.X11.XToolkit
(notebook) pwalczysko@c30f85666085:~/scripts/python$
```

**Comment by jburel** [comment]
@pwalczysko Did you start the Desktop as mentioned in the instructions?

File: docs/imagej_python.rst (hunk 2 above)
Comment (jburel) [reply docs/imagej_python.rst]: done

**Comment by pwalczysko** [comment]
> @pwalczysko Did you start the Desktop as mentioned in the instructions?

No, I did not . I find the instructions confusing, the Step by Step section basically does not have any steps at all. There are just explanations and justifications, not sure where the real workflow start, i.e. "do this do that".  See my comments above about it

**Comment by jburel** [comment]
It is more like an explanation of the script itself. It matches the step of the notebooks.

File: docs/imagej_python.rst [hunk 3]
```diff
@@ -88,7 +88,11 @@ One of the advantage of this approach is that we can load only the 2D-planes we
 The script used in this document contains an ImageJ1 macro that needs graphical user interface (GUI) elements, and thus it requires using ImageJ in GUI mode. In this GUI mode, the resulting windows content is handled.
 If you are running the example in the Docker container, 
 you will also need to start UI environment if it is not already up.
-If you have used the option 2 above, select ``desktop`` to start the UI environment. If you do not use any ImageJ1 features e.g. macro, you do **not** need the UI environment and do **not** need to set the `headless` parameter (default is `True`).
+If you have used the option 2 above, select ``desktop`` to start the UI environment.
+
+    \ |image0|
+
+If you do not use any ImageJ1 features e.g. macro, you do **not** need the UI environment and do **not** need to set the `headless` parameter (default is `True`).
```
Comment (pwalczysko) [inline docs/imagej_python.rst]:
```suggestion

```

Comment (pwalczysko) [reply docs/imagej_python.rst]: This sentence - why do we need that ? It still creates confusion.

Comment (jburel) [reply docs/imagej_python.rst]: If someone is not interested to run the macro then this is not needed.
I could write a second script and point to it.

File: docs/imagej_python.rst [hunk 4]
```diff
@@ -92,7 +92,7 @@ If you have used the option 2 above, select ``desktop`` to start the UI environm
 
     \ |image0|
 
-If you do not use any ImageJ1 features e.g. macro, you do **not** need the UI environment and do **not** need to set the `headless` parameter (default is `True`).
+If you do not use any ImageJ1 features e.g. macro, you do **not** need the UI environment.
```
Comment (pwalczysko) [inline docs/imagej_python.rst]: What if you put that sentence below the "You can now run the script."
?

For that script you need the UI environment for sure.

**Comment by pwalczysko** [comment]
@jburel understood now, thanks

```
(notebook) pwalczysko@2ef5774f772f:~/scripts/python$ ls
run_macro_python.py
(notebook) pwalczysko@2ef5774f772f:~/scripts/python$ python run_macro_python.py
Host [wss://idr.openmicroscopy.org/omero-ws]:
Image ID [1884807]:
initializing fiji...
Jun 12, 2020 10:08:19 AM java.util.prefs.FileSystemPreferences$1 run
INFO: Created user preferences directory.
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/srv/conda/vnc/Fiji.app/jars/logback-classic-1.2.3.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/srv/conda/vnc/Fiji.app/plugins/omero_ij-5.5.11-all.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [ch.qos.logback.classic.util.ContextSelectorStaticBinder]
connecting to IDR...
running the macro...
(notebook) pwalczysko@2ef5774f772f:~/scripts/python$
```

When I run this in the Terminal of the Docker, and have the Desktop open in another browser window, I can see the image being opened. This is nice.

**Comment by pwalczysko** [comment]
But locally (Option 1) I am still having probs

```
(fiji_python) pwalczysko@ls31618~/Work/omero-guide-fiji/scripts/python$ python run_macro_python.py 
Host [wss://idr.openmicroscopy.org/omero-ws]:                          
Image ID [1884807]: 
initializing fiji...
Error in `/Users/pwalczysko/miniconda3/envs/fiji_python/bin/mvn -B -f /Users/pwalczysko/.jgo/net.imglib2/imglib2-imglyb/0.4.0/pom.xml dependency:resolve': 1
Traceback (most recent call last):
  File "/Users/pwalczysko/miniconda3/envs/fiji_python/lib/python3.6/site-packages/jgo/jgo.py", line 201, in run_and_combine_outputs
    return subprocess.check_output((command,) + args, stderr=subprocess.STDOUT)
  File "/Users/pwalczysko/miniconda3/envs/fiji_python/lib/python3.6/subprocess.py", line 356, in check_output
    **kwargs).stdout
  File "/Users/pwalczysko/miniconda3/envs/fiji_python/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '('/Users/pwalczysko/miniconda3/envs/fiji_python/bin/mvn', '-B', '-f', '/Users/pwalczysko/.jgo/net.imglib2/imglib2-imglyb/0.4.0/pom.xml', 'dependency:resolve')' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "run_macro_python.py", line 90, in main
    ij = start_fiji()
  File "run_macro_python.py", line 30, in start_fiji
    ij = imagej.init('/Applications/Fiji.app', headless=False)
  File "/Users/pwalczysko/miniconda3/envs/fiji_python/lib/python3.6/site-packages/imagej/imagej.py", line 126, in init
    import imglyb
  File "/Users/pwalczysko/miniconda3/envs/fiji_python/lib/python3.6/site-packages/imglyb/__init__.py", line 38, in <module>
    config = _init_jvm_options()
  File "/Users/pwalczysko/miniconda3/envs/fiji_python/lib/python3.6/site-packages/imglyb/__init__.py", line 33, in _init_jvm_options
    import scyjava
  File "/Users/pwalczysko/miniconda3/envs/fiji_python/lib/python3.6/site-packages/scyjava/__init__.py", line 140, in <module>
    jnius = _init_jvm()
  File "/Users/pwalczysko/miniconda3/envs/fiji_python/lib/python3.6/site-packages/scyjava/__init__.py", line 126, in _init_jvm
    verbose=scyjava_config.get_verbose()
  File "/Users/pwalczysko/miniconda3/envs/fiji_python/lib/python3.6/site-packages/jgo/jgo.py", line 470, in resolve_dependencies
    mvn_out = run_and_combine_outputs(mvn, *mvn_args)
  File "/Users/pwalczysko/miniconda3/envs/fiji_python/lib/python3.6/site-packages/jgo/jgo.py", line 214, in run_and_combine_outputs
    sys.exit(e.returncode)
SystemExit: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "run_macro_python.py", line 101, in <module>
    main()
  File "run_macro_python.py", line 96, in main
    if conn:
UnboundLocalError: local variable 'conn' referenced before assignment
(fiji_python) pwalczysko@ls31618~/Work/omero-guide-fiji/scripts/python$ 

```

File: docs/imagej_python.rst [hunk 5]
```diff
@@ -9,34 +9,146 @@ This section shows how to use ImageJ as a Python library to analyze data in OMER
 Using the Python API allows us to easily load the 2D-plane we need to see or analyze.
 This is much easier that when using the Java API and Bio-Formats plugin.
```
Comment (pwalczysko) [inline docs/imagej_python.rst]:
```suggestion
This is much easier than using the Java API and Bio-Formats plugin.
```

File: docs/imagej_python.rst [hunk 6]
```diff
@@ -9,34 +9,146 @@ This section shows how to use ImageJ as a Python library to analyze data in OMER
...
-    $ conda activate fiji_python
+    $ conda activate imagej_python
 
 - Install omero-py and `pyimagej <https://pypi.org/project/pyimagej/>`_::
 
     $ conda install -c conda-forge pyimagej
 
     $ conda install -c ome omero-py
 
+Note: we have noticed few problems using that approach when running the scripts on some operating systems due to some issues with the ImageJ1-ImageJ2 bridge.
```
Comment (pwalczysko) [inline docs/imagej_python.rst]:
```suggestion
Note: we have noticed problems using the Option 1 approach when running the scripts on some operating systems due to issues with the ImageJ1-ImageJ2 bridge.
```

File: docs/imagej_python.rst [hunk 7]
```diff
@@ -9,34 +9,146 @@ This section shows how to use ImageJ as a Python library to analyze data in OMER
...
+
+- The environment with the OMERO Python bindings and a few other libraries is named ``notebook``, activate it::
+
+    $ conda activate notebook
+
+Step-by-Step
+------------
+
+The script used in this document is :download:`run_macro_python.py <../scripts/python/run_macro_python.py>`.
+One of the advantage of this approach is that we can load only the 2D-planes we wish to analyze.
```
Comment (pwalczysko) [inline docs/imagej_python.rst]:
```suggestion
One of the advantages of this approach is that we can load only the 2D-planes we wish to analyze.
```

File: docs/imagej_python.rst [hunk 8]
```diff
@@ -9,34 +9,146 @@ This section shows how to use ImageJ as a Python library to analyze data in OMER
...
+
+The script used in this document contains an ImageJ1 macro that needs graphical user interface (GUI) elements, and thus it requires using ImageJ in GUI mode. In this GUI mode, the resulting windows content is handled.
+If you are running the example in the Docker container, 
+you will also need to start UI environment if it is not already up.
+If you have used the option 2 above, select ``desktop`` to start the UI environment.
```
Comment (pwalczysko) [inline docs/imagej_python.rst]:
```suggestion
If you have used the Option 2 above, select ``desktop`` to start the UI environment.
```

File: docs/imagej_python.rst [hunk 9]
```diff
@@ -9,34 +9,146 @@ This section shows how to use ImageJ as a Python library to analyze data in OMER
...
+
+    \ |image0|
+
+
+If you are using the option 1 above, you will first need to update the script to point to your local installation of Fiji or use one of the options described in `ImageJ Tutorials <https://nbviewer.jupyter.org/github/imagej/tutorials/blob/master/notebooks/ImageJ-Tutorials-and-Demo.ipynb>`_.
```
Comment (pwalczysko) [inline docs/imagej_python.rst]:
```suggestion
If you are using the Option 1 above, you will first need to update the script to point to your local installation of Fiji or use one of the options described in `ImageJ Tutorials <https://nbviewer.jupyter.org/github/imagej/tutorials/blob/master/notebooks/ImageJ-Tutorials-and-Demo.ipynb>`_.
```

**Comment by pwalczysko** [comment]
Couple of minor formatting suggestions, otherwise good to go.

**Comment by jburel** [comment]
Thanks @pwalczysko

**Pull request merged**